lightning/restore/restore.go

```diff
@@ -1344,19 +1360,37 @@ func setSessionConcurrencyVars(ctx context.Context, db *sql.DB, dsn config.DBSto
 // DoChecksum do checksum for tables.
 // table should be in <db>.<table>, format.  e.g. foo.bar
 func DoChecksum(ctx context.Context, db *sql.DB, table string) (*RemoteChecksum, error) {
-	ori, err := increaseGCLifeTime(ctx, db)
-	if err != nil {
-		return nil, errors.Trace(err)
+	var err error
+	helper, ok := ctx.Value(&gcLifeTimeKey).(gcLifeTimeHelper)
+	if !ok {
+		return nil, errors.Errorf("No gcLifeTimeHelper found in context, check context initialization")
+	}
+
+	helper.increaseGCLock.Lock()
+	*(helper.runningJobs) += 1
+	if *(helper.runningJobs) == 1 {
+		err = increaseGCLifeTime(ctx, db)
+		if err != nil {
+			helper.increaseGCLock.Unlock()
+			return nil, errors.Trace(err)
+		}
 	}
+	helper.increaseGCLock.Unlock()
+
 	// set it back finally
 	defer func() {
-		err := UpdateGCLifeTime(ctx, db, ori)
-		if err != nil {
-			query := fmt.Sprintf("UPDATE mysql.tidb SET VARIABLE_VALUE = '%s' WHERE VARIABLE_NAME = 'tikv_gc_life_time'", ori)
-			log.L().Warn("revert GC lifetime failed, please reset the GC lifetime manually after Lightning completed",
-				zap.String("query", query),
-				log.ShortError(err),
-			)
+		if atomic.AddInt32(helper.runningJobs, -1) == 0 {
```

Comment:
Can this scenario happen, the condition `atomic.AddInt32(helper.runningJobs, -1) == 0` in this line is true, but before L1383 executes, another RetstoreTable routine starts and increases `helper.runningJobs` by 1 ?

Reply:
You're right.
In fact another RestoreTable routine has its own `helper.runningJobs`, 
[L507](https://github.com/pingcap/tidb-lightning/pull/220/files#diff-9697521ce51bfb4c9ef44a387e73c780R507). But two tightly running RestoreTable routines will have undetermined result.

Now I want to make `helper.runningJobs` point to one global integer which represents only one TiDB, and use same lock not atomic operations. Hence `gcLifeTimeLock` and `runningJobs` become file-global. What do you advise?

Reply:
I'm not familiar with the `server mode` feature, but if the task is running one by one in the server mode, not in a concurrent way, a global file-level lock for `runningJobs` is fine.